sys/net/gnrc/network_layer/ndp2/gnrc_ndp2.c

```diff
@@ -0,0 +1,567 @@
+#include "debug.h"
+
+gnrc_pktsnip_t *gnrc_ndp2_nbr_sol_build(const ipv6_addr_t *tgt,
+                                        gnrc_pktsnip_t *options)
+{
+    gnrc_pktsnip_t *pkt;
+
+    assert((tgt != NULL) && !ipv6_addr_is_multicast(tgt));
+    DEBUG("ndp2: building neighbor solicitation message\n");
+    pkt = gnrc_icmpv6_build(options, ICMPV6_NBR_SOL, 0, sizeof(ndp_nbr_sol_t));
+    if (pkt != NULL) {
+        ndp_nbr_sol_t *nbr_sol = pkt->data;
+        nbr_sol->resv.u32 = 0;
+        nbr_sol->tgt.u64[0].u64 = tgt->u64[0].u64;
+        nbr_sol->tgt.u64[1].u64 = tgt->u64[1].u64;
+    }
+    return pkt;
+}
+
+gnrc_pktsnip_t *gnrc_ndp2_nbr_adv_build(const ipv6_addr_t *tgt, uint8_t flags,
+                                        gnrc_pktsnip_t *options)
+{
+    gnrc_pktsnip_t *pkt;
+
+    assert((tgt != NULL) && !ipv6_addr_is_multicast(tgt));
+    DEBUG("ndp2: building neighbor advertisement message\n");
+    pkt = gnrc_icmpv6_build(options, ICMPV6_NBR_ADV, 0, sizeof(ndp_nbr_adv_t));
+    if (pkt != NULL) {
+        ndp_nbr_adv_t *nbr_adv = pkt->data;
+        nbr_adv->flags = (flags & NDP_NBR_ADV_FLAGS_MASK);
+        nbr_adv->resv[0] = nbr_adv->resv[1] = nbr_adv->resv[2] = 0;
+        nbr_adv->tgt.u64[0].u64 = tgt->u64[0].u64;
+        nbr_adv->tgt.u64[1].u64 = tgt->u64[1].u64;
+    }
+    return pkt;
+}
+
+gnrc_pktsnip_t *gnrc_ndp2_rtr_sol_build(gnrc_pktsnip_t *options)
+{
+    gnrc_pktsnip_t *pkt;
+
+    DEBUG("ndp2: building router solicitation message\n");
+    pkt = gnrc_icmpv6_build(options, ICMPV6_RTR_SOL, 0, sizeof(ndp_rtr_sol_t));
+    if (pkt != NULL) {
+        ndp_rtr_sol_t *rtr_sol = pkt->data;
+        rtr_sol->resv.u32 = 0;
+    }
+    return pkt;
+}
+
+gnrc_pktsnip_t *gnrc_ndp2_rtr_adv_build(uint8_t cur_hl, uint8_t flags,
+                                        uint16_t ltime, uint32_t reach_time,
+                                        uint32_t retrans_timer,
+                                        gnrc_pktsnip_t *options)
+{
+    gnrc_pktsnip_t *pkt;
+
+    DEBUG("ndp2: building router advertisement message\n");
+    pkt = gnrc_icmpv6_build(options, ICMPV6_RTR_ADV, 0, sizeof(ndp_rtr_adv_t));
+    if (pkt != NULL) {
+        ndp_rtr_adv_t *rtr_adv = pkt->data;
+        rtr_adv->cur_hl = cur_hl;
+        rtr_adv->flags = (flags & NDP_RTR_ADV_FLAGS_MASK);
+        rtr_adv->ltime = byteorder_htons(ltime);
+        rtr_adv->reach_time = byteorder_htonl(reach_time);
+        rtr_adv->retrans_timer = byteorder_htonl(retrans_timer);
+    }
+    return pkt;
+}
+
+static inline size_t _ceil8(uint8_t length)
+{
+    /* NDP options use units of 8 byte for there length field, so round up */
+    return (length + 7U) & 0xf8U;
+}
+
+gnrc_pktsnip_t *gnrc_ndp2_opt_build(uint8_t type, size_t size,
+                                    gnrc_pktsnip_t *next)
+{
+    ndp_opt_t *opt;
+    gnrc_pktsnip_t *pkt = gnrc_pktbuf_add(next, NULL, _ceil8(size),
+                                          GNRC_NETTYPE_UNDEF);
+
+    if (pkt == NULL) {
```

Comment:
I actually try to avoid early exits since they tend to produce more machine code.